Refs #4478 - prebuild localized apipie cache

mbacovsky opened the pull request:
Requires https://github.com/theforeman/foreman/pull/1428 to get merged.
Documentation for resources is built during build time. Docs for Index is build in posttrans phase as we need foreman plugins to be present.

domcleal commented:
Just a reminder, this needs finishing off for 1.8 (for both RPMs and debs) else I'm going to have to revert the change to Foreman core - the apipie:cache performance is simply dreadful at the moment out of the box.

mbacovsky commented:
I rebased the PR. 
I changed the default options in foreman_precompile_plugin macro, I tried to avoid that, but this way it seems cleaner. The affected packages were updated in separate commits. I'll have to add the package names to the commit messages.

Rubygem foreman bootdisk updated and scratch builds of all the related packages are still WIP

lzap commented:
@ohadlevy you asked on Discovery API docs generation yesterday - once this is merged we should have it.

mbacovsky commented:
I updated the commit messages and added foreman-bootdisk.

domcleal commented on foreman/foreman.spec:
```diff
@@ -570,6 +588,8 @@ rm -rf %{buildroot}
 %config %{_sysconfdir}/cron.d/%{name}
 %{_sysconfdir}/rpm/macros.%{name}
+%{_sysconfdir}/rpm/macros.%{name}-apipie
+%{_sysconfdir}/rpm/macros.%{name}-assets
```
Should be removed from -assets' %files

mbacovsky replied:
You mean the %foreman_precompile_plugin macro should be romoved from -assets file? Should I move it to -plugin?

domcleal replied:
Sorry, I meant that if you're adding the -assets macro file to 'foreman' then it can be removed from 'foreman-assets' so it's not in both (though harmless).

I do kind of like the idea of a -plugin or -build subpackage though, which could depend on foreman-sqlite and ship these macro files.  Or we repurpose foreman-assets for this too?

mbacovsky replied:
Ahh I see now. I'd prefer to merge assets and apipie macros to macros.foreman-plugin and ship that in plugin subpackage with the sqlite dependency. I'll update the PR that way if there are no objections @domcleal.

domcleal replied:
Yep, I like that too.

domcleal replied:
Oh, don't forget to update comps/ when adding a sub-package, so it's shipped.

domcleal commented on foreman/foreman.spec:
```diff
@@ -540,10 +544,24 @@ cp %%{buildroot}%%{%{name}_bundlerd_dir}/%%{gem_name}.rb ./bundler.d/%%{gem_name
+
+%%%{name}_apipie_cache_plugin_post \\
++cp -r %%{foreman_apipie_cache_foreman}/* %%{foreman_dir}/public/apipie-cache/ \\
++chown -R %{name}.%{name} %%{foreman_dir}/public/apipie-cache
```
additional leading "+" on these lines, that doesn't look right to me?

mbacovsky replied:
fixed

domcleal commented on foreman/foreman.spec:
```diff
@@ -540,10 +544,24 @@ cp %%{buildroot}%%{%{name}_bundlerd_dir}/%%{gem_name}.rb ./bundler.d/%%{gem_name
 export BUNDLER_EXT_NOSTRICT=1 \\
-/usr/bin/%%{?scl:%%{scl}-}rake %%{-r*}%%{!?-r:plugin:assets:precompile[%%{-n*}%%{!?-n:%%{gem_name}}]} RAILS_ENV=production --trace \\
+%%{?-s:/usr/bin/%%{?scl:%%{scl}-}rake %%{-r*}%%{!?-r:plugin:assets:precompile[%%{-n*}%%{!?-n:%%{gem_name}}]} RAILS_ENV=production --trace} \\
+%%{?-a:/usr/bin/%%{?scl:%%{scl}-}rake plugin:apipie:cache[%%{gem_name}] RAILS_ENV=development cache_part=resources OUT=%%{buildroot}%%{gem_instdir}/public/apipie-cache/plugin/%%{gem_name} #gem_instdir}/public/apipie-cache/plugin/%%{gem_name} --trace} \\
```
is there a comment halfway through this line?

mbacovsky replied:
Oh, fixed

domcleal commented on foreman/foreman.spec:
```diff
@@ -621,6 +641,10 @@ if [ ! -e %{_datadir}/%{name}/config/initializers/encryption_key.rb -a \
   ln -s %{_sysconfdir}/%{name}/encryption_key.rb %{_datadir}/%{name}/config/initializers/
 fi
 
+# copy pregenerated apipie cache files
```
Shouldn't this be in the build?

mbacovsky replied:
This is mainly for case of Foreman update. This code is responsible for copying of the plugin's pre-built caches to the main `apipie-cache` directory. It can not happen during the build time because of conflict (resource with same name in core and plugin).

domcleal replied:
Interesting, so what wins in that case?  It looks like it would be the most recently installed package, rather than anything deterministic - e.g. the latest Foreman version, or the latest plugin version?

domcleal replied:
Another thought, if we want Foreman core to always win, we could have plugins %exclude the conflicting files in their file list.

mbacovsky replied:
The idea was that plugin provides docs for resource only if it overrides the core. As it depednds on the core it should be loaded later and win. It is true that with multiple overriding plugins there no rule (is this situation covered on route level?) There might be also  problem with core update.

We can move the copying to the rake task perhaps. That would solve the problem with update and we have access to Rails there if we have to figure out the plugin ordering. What do you think?

domcleal replied:
I feared you'd say it was that way around :)  Yeah, we need to sort this in core or apipie-rails then, rather than packaging, as we need to be able to handle core-without-plugin updates reliably.  I've no idea about routes, probably depends on how the plugin is written.

One idea before we go the copying-in-rake route.. today, we serve up precompiled assets from multiple plugin directories: https://github.com/theforeman/foreman/blob/1.7.1/config/environments/production.rb#L140-L147

Would it be possible to serve up static apipie files in the same way - if we can order the plugins above core?  Then we could build the index of all the plugins, and even avoid copying them out of the plugin package.

If not, then copying in the rake task that builds the index is the next best thing to do.  We would need to support a directory location outside of the gem installation root due to support Debian packaging (same reason as https://github.com/theforeman/foreman/commit/d43379f323c507d19a2e8f051cd2e767be84125e), i.e. stick with /usr/share/foreman/public/apipie-cache/plugin/%{plugin-name}.

domcleal replied:
I'm just going to test out that idea..

domcleal replied:
That idea doesn't work, as we're actually going through an apipie-rails controller rather than statically serving files, so I think we'll have to copy and combine files in the rake task into the cache dir.

domcleal commented on rubygem-foreman_bootdisk/rubygem-foreman_bootdisk.spec:
```diff
@@ -42,6 +42,7 @@ BuildRequires: %{?scl_prefix}ruby(release)
 %else
 BuildRequires: %{?scl_prefix}ruby(abi) >= %{rubyabi}
 %endif
+BuildRequires:   foreman >= 1.8.0
```
indentation's off

domcleal commented on rubygem-foreman_bootdisk/rubygem-foreman_bootdisk.spec:
```diff
@@ -82,16 +83,20 @@ cat <<GEMFILE > %{buildroot}%{foreman_bundlerd_dir}/%{gem_name}.rb
 gem '%{gem_name}'
 GEMFILE
 
+%foreman_precompile_plugin -a
```
This is failing, as the rake task's trying to initialise the database, and we don't have foreman-sqlite as a build dep (thankfully?).

<pre>
+ /usr/bin/ruby193-rake 'plugin:apipie:cache[foreman_bootdisk]' RAILS_ENV=development cache_part=resources OUT=/builddir/build/BUILDROOT/ruby193-rubygem-foreman_bootdisk-4.0.2-1.git.188.6c277bf.el7.centos.x86_64/opt/rh/ruby193/root/usr/share/gems/gems/foreman_bootdisk    -4.0.2/public/apipie-cache/plugin/foreman_bootdisk
Rails Error: Unable to access log file. Please ensure that /builddir/build/BUILD/rubygem-foreman_bootdisk-4.0.2/usr/share/foreman/log/development.log exists and is chmod 0666. The log level has been raised to WARN and the output directed to STDERR until the problem is     fixed.
rake aborted!
Please install the sqlite3 adapter: `gem install activerecord-sqlite3-adapter` (Could not find sqlite3 (~> 1.3.5) amongst [actionmailer-3.2.8, actionpack-3.2.8, activemodel-3.2.8, activerecord-3.2.8, activeresource-3.2.8, activesupport-3.2.8, ancestry-2.0.0, apipie-ra    ils-0.2.6, arel-3.0.2, audited-3.0.0, audited-activerecord-3.0.0, bigdecimal-1.1.0, builder-3.0.0, bundler-1.3.5, bundler_ext-0.3.0, coffee-rails-3.2.2, coffee-script-2.2.0, coffee-script-source-1.3.3, deep_cloneable-2.0.2, diff-lcs-1.1.3, erubis-2.7.0, execjs-1.4.0,     fast_gettext-0.9.2, foreigner-1.7.1, foreman_bootdisk-4.0.2, friendly_id-4.0.10.1, gettext_i18n_rails-1.2.0, hike-1.2.1, i18n-0.6.0, io-console-0.3, journey-1.0.4, json-1.5.5, ldap_fluff-0.3.3, mail-2.4.4, mime-types-1.19, minitest-2.5.1, multi_json-1.10.1, net-http-p    ersistent-2.7, net-ldap-0.10.0, netrc-0.7.7, oauth-0.4.7, polyglot-0.3.3, rabl-0.11.4, rack-1.4.1, rack-cache-1.2, rack-ssl-1.3.2, rack-test-0.6.1, rails-3.2.8, railties-3.2.8, rake-0.9.2.2, rdoc-3.9.5, rest-client-1.6.7, ruby2ruby-2.1.3, ruby_parser-3.6.3, safemode-1    .2.2, scoped_search-2.7.1, secure_headers-1.4.1, sexp_processor-4.4.4, sprockets-2.8.2, thor-0.18.1, tilt-1.3.3, treetop-1.4.10, turbolinks-2.5.3, tzinfo-0.3.33, validates_lengths_from_database-0.4.0, will_paginate-3.0.2])
Tasks: TOP => plugin:apipie:cache
(See full trace by running task with --trace)
</pre>

Is there any way we can fix the rake task not to initialise the database connection, or is it necessary?

domcleal commented on foreman/foreman.spec:
```diff
@@ -540,10 +544,24 @@ cp %%{buildroot}%%{%{name}_bundlerd_dir}/%%{gem_name}.rb ./bundler.d/%%{gem_name
+%%%{name}_apipie_cache_foreman %%{foreman_dir}/public/apipie-cache/plugin/%%{gem_name}
+
+%%%{name}_apipie_cache_plugin_post \\
++cp -r %%{foreman_apipie_cache_foreman}/* %%{foreman_dir}/public/apipie-cache/ \\
```
similar query to below about the copy/build stage, why are there two copies of this?  It's also visible in %files in bootdisk for instance, which seems strange to me.

domcleal replied:
If there really must be two copies for some reason, and we can't remove it or symlink them, then we should copy them at build time and have the package ship them both, not copy it in %post.  This will ensure that file ownership, file cleanup, "rpm -V" etc works, and the RPM should compress away the second copy.

mbacovsky replied:
For a plugin there is one copy of the prebuilt cache in `%{gem_instdir}/public/apipie-cache/plugin/%{gem_name}` and it is linked to the `%{foreman_dir}/public/apipie-cache/plugin/%%{gem_name}`.
The plugin can contailn the same resources as the foreman core (e.g. environments in Foreman and Katello) so it is not possible to link it to `%{foreman_dir}/public/apipie-cache/` directly in build time as it would cause conflicts. Therefore we copy it in post and overwriting the duplicites.

I agree it would be better to have just one copy in the package. And I would prefere to drop the copy in the `gem_instdir`. Would that be okay?

mbacovsky commented:
Part of the issues was fixed. I'll have to investigate the issue with bootdisk. The rake task needs to load rails environment and that requires DB. For example for Katello I was unable to avoid that. I'll have to discuss it with @iNecas.

iNecas commented:
Just to share our experience with trying to avoid the database requirement when building the apipie documentation, trying to workaround the Rails database needs lead to big pain every time something changed (either in our code or in Rails). The only other thinkable solution would be to not use the model classes inside the apipie definitions, but that's not very convenient either.

mbacovsky commented:
@domcleal I tried to rebuild rubygem-foreman_bootdisk with updated foreman and it worked. The result rpm contains the prebuilt cache. The build log is here http://fpaste.org/171401/42167563/ and the Foreman scratch build can be found here http://koji.katello.org/koji/taskinfo?taskID=212128 (both on RHEL 7) 
Foreman_bootdisk scratch: https://mbacovsk.fedorapeople.org/packages/ruby193-rubygem-foreman_bootdisk-4.0.2-1.el6.noarch.rpm, https://mbacovsk.fedorapeople.org/packages/ruby193-rubygem-foreman_bootdisk-4.0.2-1.el6.src.rpm

mbacovsky commented:
The foreman package has been rebased

domcleal commented:
Regarding the sqlite dep, that's OK, we'll live with it.

I think the build worked for you @mbacovsky because you may have had ruby193-rubygem-sqlite installed on your build host.  It failed for me in a clean mock environment.

mock is very easy to set up - install it, then copy one of the configs from #389 to /etc/mock, e.g. /etc/mock/el7-scl-foreman.cfg (that's how I have mine).  If you then generate SRPMs (`rpmbuild -bs`) you can do a chain build to test the foreman build, followed by the plugin that depends on the macros from the first:

    mockchain -r el7-scl-foreman -m --yum foreman-1.8.0-0.develop.201501141458gitd43379f.el7.src.rpm rubygem-foreman_bootdisk-4.0.2-1.git.188.6c277bf.fc21.src.rpm

As mentioned in the inline comment above, we could perhaps put the foreman-sqlite dependency on foreman-assets and keep macros in that file, or put it all in foreman-plugin/foreman-build and have plugins depend on that instead?

mbacovsky commented:
The PR was rebased and updated. I've added the changes in separate TMP commits and will squash them later if they are found okay. Hopefully it makes the review of the changes easier.

- moved the copying of the cache to the rake task so it requires https://github.com/theforeman/foreman/pull/2104
- moved the macros and sqlite dependency to the foreman-plugin subpacakge and updated the dependences.
- removed the core apipie cache from public/apipie-cache/plugin/foreman as it seems unnecessary with current changes

@domcleal thatnks for the mocchain howto, it helped with the testbuilds significantly. Currently there are problems only with bastion and discovery. I'm investigating that, but it seems to be some local issue 
unrelated to my changes. Everything else builds fine.

mbacovsky commented:
Updated, discovery was missing new build dep and bastion builds okay after rebase. So all packages addressed with this PR build okay now.

domcleal commented on foreman/foreman.spec:
```diff
@@ -656,7 +678,7 @@ exit 0
 %{foreman_rake} db:migrate db:compute_resources:encrypt >> %{_localstatedir}/log/%{name}/db_migrate.log 2>&1 || :
 %{foreman_rake} db:seed >> %{_localstatedir}/log/%{name}/db_seed.log 2>&1 || :
-%{foreman_rake} apipie:cache >> %{_localstatedir}/log/%{name}/apipie_cache.log 2>&1 || :
+%{foreman_rake} apipie:cache:index >> %{_localstatedir}/log/%{name}/apipie_cache.log 2>&1 || :
```
The index call will be needed in the plugin %posttrans too, in case they're installed after Foreman.  Maybe update `%%%{name}_apipie_cache` in foreman.spec to do `:index` and use that?

mbacovsky replied:
I'm curious what happens when foreman and 3 plugins all with same :index in posttrans are installed. Will it trigger the :index four times?

domcleal replied:
Yep.. I think that's unavoidable for now.  Apart from the rake/app load speed, it's reasonably fast.

mbacovsky replied:
Okay, I'll fix that.

domcleal commented on foreman/foreman.spec:
```diff
@@ -325,7 +325,19 @@ Meta package to install asset pipeline support.
+
+%package plugin
+Summary: Foreman plugin support
+Group: Applications/system
```
Nitpick: Development/Libraries perhaps?

domcleal commented on rubygem-foreman_bootdisk/rubygem-foreman_bootdisk.spec:
```diff
@@ -42,6 +42,8 @@ BuildRequires: %{?scl_prefix}ruby(release)
 %else
 BuildRequires: %{?scl_prefix}ruby(abi) >= %{rubyabi}
 %endif
+BuildRequires: foreman >= 1.8.0
```
unnecessary?

mbacovsky replied:
Yes, indeed

domcleal commented on rubygem-foreman_discovery/rubygem-foreman_discovery.spec:
```diff
@@ -49,6 +49,9 @@ BuildRequires: %{?scl_prefix}ruby(release)
 %else
 BuildRequires: %{?scl_prefix}ruby(abi) >= %{rubyabi}
 %endif
+BuildRequires: foreman >= 1.8.0
```
unnecessary?

domcleal commented:
I like the look of this a lot, good work!  Once the %posttrans bit is fixed, I'll take a look at, and test the built packages.

mbacovsky commented:
Updated. I moved %foreman_apipie_cache to the foreman-plugin and let it build just the index. Posttrans in bootdisk and discovery were updated. The index is being rebuild multiple times when installing more packages with cache, but one run takes about 26sec on my setup so it could be acceptable.

@domcleal, please let me know when you want me to squash the commits.

mbacovsky commented:
This patch to foreman_bootdisk https://github.com/theforeman/foreman_bootdisk/pull/6 needs to be merged before the changes of packaging.

domcleal commented on foreman/foreman.spec:
```diff
@@ -424,6 +436,8 @@ cp config/settings.yaml.example config/settings.yaml
 export BUNDLER_EXT_NOSTRICT=1
 export BUNDLER_EXT_GROUPS="default assets"
 %{scl_rake} assets:precompile:all RAILS_ENV=production --trace
+%{scl_rake} db:migrate RAILS_ENV=production --trace
```
I think this also needs to be in the RPM macros so plugins also migrate during build.  The foreman_discovery build fails right now as it can't open the DB while running plugin:apipie:cache.

mbacovsky replied:
@domcleal, I've added the migration to the macro. While building discovery the migration finish successfully, cache is generated, but the build fails due to some other error. Investigating...
```
 error: File not found: /builddir/build/BUILDROOT/ruby193-rubygem-foreman_discovery-2.0.0-1.el7.centos.x86_64/opt/rh/ruby193/root/usr/share/gems/gems/foreman_discovery-2.0.0/public
```

mbacovsky replied:
Just for the record, I was building discovery from 2.0 stable branch and overlooked that apipie initialization patch was removed from the stable branch (remains in the develop)

mbacovsky commented:
Updated, rebased. 
- migration was added
- in discovery I replaced bundler.d creation with macro

- Does it make sense to run db:seed as well?  I think just Katello plugin needs that though.
- There are some other plugin related macros like %foreman_bundlerd_file left in the macros.foreman. Does it make sense to move them to foreman-plugin?

domcleal commented:
> Does it make sense to run db:seed as well? I think just Katello plugin needs that though.

Sure.

> There are some other plugin related macros like %foreman_bundlerd_file left in the macros.foreman. Does it make sense to move them to foreman-plugin?

Yeah, good idea.

mbacovsky commented:
@domcleal, updated

domcleal commented:
I'm struggling to get the db:migrate to work with foreman_bootdisk (.gem generated from master branch) as ./usr/share/foreman/db is a symlink to /var/lib/foreman/db in the build environment, and it doesn't have permission to create the sqlite database.

I'm trying to copy /var/lib/foreman/db to ./var and then change the symlink, but haven't quite got it to work yet.  I wonder if you encountered the same?

mbacovsky commented:
@domcleal, not sure if I understand when the problem occures. My chain build is not failing and the db:merge proceeds without complaints. Here is a link to a build log (a couple days old Foreman develop and fresh foreman_bootdisk gem generated from master branch) http://fpaste.org/184611/39405142/

domcleal commented:
Thanks, I'll dig into it a bit more.  Here's the build log for reference: http://paste.fedoraproject.org/184624/23740802/

mbacovsky commented:
@domcleal, I've updated again. I was able to build foreman and foreman_bootdisk as non-root with these specs. There are wierd errors in the build log while db:seed runs. The seed seems to finish okay though. Running it for second time does not produce the errors anymore. Investigating...

mbacovsky commented:
@domcleal, updated. Even with @iNecas's help we didn't found out in reasonable time why the roles/permissions seed is failing when ran as non-root. But it seems that the seeds are not necessary to build the cache so I removed them. It seems I was mistaken and even Katello does not need preseeded database to generate the apipie cache, so it should be safe. 
My test builds are looking good now.

domcleal commented:
@mbacovsky great, thanks.  Can you just do a git rebase please as there's a merge error?

mbacovsky commented:
Rebased

domcleal commented:
Merged as 53c3308..9e446af.  Nice one, thanks @mbacovsky!

mbacovsky commented:
Thanks a 1e6 for your patience, @domcleal !!!  I owe you some well deserved :beers:

lzap commented:
Guys, this patch changed dependency from >= 1.7 to >= 1.8 which is not yet true and plugin won't load.

lzap commented on rubygem-foreman_discovery/rubygem-foreman_discovery.spec:
```diff
@@ -34,7 +34,7 @@ License:    GPLv3
 Source0:    http://rubygems.org/downloads/%{gem_name}-%{version}%{?prever}.gem
 
-Requires:   foreman >= 1.7.0
+Requires:   foreman >= 1.8.0
```
I mean for Discovery plugin...

domcleal commented:
@lzap it is true, as the changes to the spec file require 1.8.0.  We'll simply build a second 1.7 compatible version on rpm/1.7 in the same way I did for foreman_docker.
